project_euler/problem_188/sol1.py

```diff
@@ -0,0 +1,74 @@
+"""
+The hyperexponentiation of a number
+Problem 188
+
+The hyperexponentiation or tetration of a number a by a positive integer b,
+denoted by a↑↑b or b^a, is recursively defined by:
+
+a↑↑1 = a,
+a↑↑(k+1) = a(a↑↑k).
+
+Thus we have e.g. 3↑↑2 = 3^3 = 27, hence 3↑↑3 = 3^27 = 7625597484987 and
+3↑↑4 is roughly 103.6383346400240996*10^12.
+
+Find the last 8 digits of 1777↑↑1855.
+
+References:
+    - The original Project Euler project page:
+      https://projecteuler.net/problem=188
+    - Wikipedia article about Hyperexponentiation, aka. Tetration
+      https://en.wikipedia.org/wiki/Tetration
+"""
+
+
+# small helper function for modular exponentiation
+def _modexpt(base: int, exponent: int, modulo_value: int) -> int:
+    """
+    Returns the modular exponentiation, that is the value
+    of `base ** exponent % modulo_value`, without calculating
+    the actual number.
+    >>> _modexpt(2, 4, 10)
+    6
+    >>> _modexpt(2, 1024, 100)
+    16
+    >>> _modexpt(13, 65535, 7)
+    6
+    """
+
+    if exponent == 1:
+        return base
+    if exponent % 2 == 0:
+        x = _modexpt(base, exponent / 2, modulo_value) % modulo_value
+        return (x * x) % modulo_value
+    else:
+        return (base * _modexpt(base, exponent - 1, modulo_value)) % modulo_value
+
+
+def solution(base: int = 1777, height: int = 1855, digits: int = 8) -> int:
+    """
+    Returns the last 8 digits of the hyperexponentiation of base by
+    height, i.e. the number base↑↑height:
+
+    >>> solution()
+    95962097
```

Comment:
You don't need to test the answer here as it is being tested by this [script](https://github.com/TheAlgorithms/Python/blob/master/project_euler/validate_solutions.py). Provide `doctest` to test the function and not the answer.

Reply:
I don't think I understand. Do you want me to remove that one test and leave the other two in place? Isn't the answer to the problem also a test of the function?

Reply:
I am going to write guidelines regarding code style for Project Euler to maintain consistency throughout the directory. What I'm proposing here is that the above-linked script is already checking whether the answer is correct or not. Now, if we include it in `doctest` again then Travis CI will run that test and so we will have duplicate tests. To remove that I'm proposing to not include the test for the answer but include tests for the function. Check whether the function is working appropriately by passing arguments other than the default ones (which gives us the answer).

Example: If `solution(n = 100)` gives the answer to the problem, test the function for `solution(n = 5)`, `solution(n = 0)`, `solution(n = 20)` and so on. Check whether the answer to those calls matches what you expect from the function.